### ISSUE TITLE
Nvidia drivers fail to install

### ISSUE THREAD

**Issue by cakmakb** [NixOS/nixpkgs#205125]
### Steps To Reproduce
- Add this into configuration.nix:
```console
nixpkgs.config.allowUnfree = true;
services.xserver.videoDrivers = [ "nvidia" ];
hardware.opengl.enable = true;
hardware.nvidia.package = config.boot.kernelPackages.nvidiaPackages.production;
```
- execute: sudo nixos-rebuild switch 

### Expected behavior
Successfull installation of nvidia drivers

### Build log
[Link](https://gist.github.com/cakmakb/cc5acbd09f9a6550eade9f9364c118a5)

### Hardware
- HP Zbook G6 17
- CPU: Intel Xeon e-2286m
- GPU: Nvidia Quadro RTX 5000

### Additional context
- Tested with hybrid graphics on/off
- Tested with Kernels 4.19, 5.15, 6.0
- Nvidia drivers 515-520 worked before on Pop_Os!/EndeavourOS on Kernels 5.1x, 5.19, 6.0
- I tried stable, production, latest, beta, vulkan_beta and even legacy_470 without success. The examples here used production. 
- Tested different USB drives and even a DVD as installation media.
- Booted into a live session from the install media, inserted the lines above into configuration.nix and tried to rebuild. On an older Laptop with a Nvidia card the driver is installed successfully, but not on this machine.
- If I boot with kernel parameter "acpi=off" the installation is successful!
- After installation with acpi=off and removal of the kernel parameter the nvidia driver is in a working condition

### Metadata
```console
[cakmakb@hpZbook17:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.81, NixOS, 22.11 (Raccoon), 22.11.691.5528350186a`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.11.0`
 - channels(cakmakb): `""`
 - channels(root): `"nixos-22.11"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by NickCao** [comment]
The build process should not be affected by your graphic settings and kernel parameters as it is executed in the nix sandbox, the phenomenon you described does not make sense to me.

**Comment by NickCao** [comment]
`linuxPackages_5_15.nvidiaPackages.production` builds for me on nixpkgs commit `5528350186a`, failed to reproduce the build failure.

**Comment by cakmakb** [comment]
Thanks for your reply! 

It doesn't make sense for me either but thats just what happens. If I boot the same installation media on my old Schenker Laptop copy

```console
# NVIDIA drivers are unfree.
  nixpkgs.config.allowUnfree = true;

  services.xserver.videoDrivers = [ "nvidia" ];
  hardware.opengl.enable = true;
```

from the [Wiki](https://nixos.wiki/wiki/Nvidia) it works. If I do exactly the same on the HP it fails.

**Comment by NickCao** [comment]
A followup to this issue, `/proc/cmdline` is actually available is the sandbox, introducing some impurity into the sandbox (which I would consider as a nix bug)

**Comment by cakmakb** [comment]
Thanks for keeping this issue in mind!

First: Is there anything I can do to provide more Information?

I'm just trying to understand: I've read `/proc/cmdline` contains the kernel parameters provided to the kernel on boot. So this should be the same string I added my `acpi=off` by pressing e in the boot menu, right? How can this cause issues later on if the system is up and running / while building something? Except for the hashes it contains the same on my other machine that works: 

`initrd=\efi\nixos\96j5hm4hg0q35cah13dvjqgkhwm01i2h-initrd-linux-5.15.83-initrd.efi init=/nix/store/cjpfww41qf9b6z27fyskfc637dp2vxf7-nixos-system-hpZbook17-22.11.1008.0938d73bb14/init loglevel=4
`

I had to `nixos-rebuild switch` today since I installed new software. I had a couple of minutes so I also added a `--upgrade` and it failed when building the nvidia drivers. I could resolve it again by temporarily adding `acpi=off` as a kernel parameter. I am now running kernel 5.15-83 and nvidia 520.56.06 successfully after a reboot. Btw. I had no issues with vmware-workstation. I can remember I had to build kernel modules on other distros for vmware in the past.

A couple of days ago I had some spare time and I had a look into the build log I posted. Some errors remind me of issues I had a long time ago when I tried to compile something with a wrong compiler (or wrong version of a compiler). The first error for example is:  `/nix/store/95dsmwshpvw8lc5gbjxfl7qivdcsg9rp-linux-5.15.81-dev/lib/modules/5.15.81/source/include/asm-generic/percpu.h:2: error: unterminated #ifndef`. I looked into the source file, yes there is no #ifndef. But if this is error worthy nobody should be able to build it. Could this be related to the multiple entries `cc1: some warnings being treated as errors`? Maybe some missing parameters for compilation? I mean even if it builds successfully there must be 100s or 1000s of warnings I can see flying by. If these are treated as errors of course the build is going to fail.

Anyway thanks again for your reply. If it's holiday season for you take some time off, this is not a deal breaker for me ;-)

**Comment by NickCao** [comment]
> I'm just trying to understand: I've read /proc/cmdline contains the kernel parameters provided to the kernel on boot. So this should be the same string I added my acpi=off by pressing e in the boot menu, right? How can this cause issues later on if the system is up and running / while building something? 

Yes, `/proc/cmdline` contains exactly the params you passed to kernel on boot (from your bootloader). The build process of some kernel modules (namely the nvidia ones) would read `/proc/cmdline` (a bad practice!) to gain information about the running system and adjust itself accordingly, for example it could disable building acpi support when seeing `acpi=off`, causing the build to behave indeterministically.

> error: unterminated #ifndef

This is indeed error worthy (it is an error, by definition). Some projects may treat certain types of warning as errors while leaving the rest as is, regarding the logs you've skimmed.
